File: webmvp/src/lib/songSearchText.test.ts
import { describe, expect, it } from "vitest";

import {
  buildSongSearchText,
  flattenSectionsLyrics,
  SONG_INDEX_SEARCH_TEXT_MAX,
} from "@/lib/songSearchText";
import type { Section } from "@/lib/types";

const sampleSections: Section[] = [
  {
    label: "Verse 1",
    lines: [
      { lyrics: "Amazing grace, how sweet the sound", chords: [] },
      { lyrics: "That saved a wretch like me", chords: [] },
    ],
  },
];

describe("flattenSectionsLyrics", () => {
  it("returns all lyric lines in order", () => {
    expect(flattenSectionsLyrics(sampleSections)).toEqual([
      "Amazing grace, how sweet the sound",
      "That saved a wretch like me",
    ]);
  });
});

describe("buildSongSearchText", () => {
  it("includes title, artist, tags, and lyrics", () => {
    const text = buildSongSearchText({
      title: "Amazing Grace",
      artist: "John Newton",
      tags: ["lang:english", "worship"],
      sections: sampleSections,
    });

    expect(text).toContain("amazing grace");
    expect(text).toContain("john newton");
    expect(text).toContain("lang:english");
    expect(text).toContain("sweet the sound");
    expect(text).toContain("wretch like me");
  });

  it("normalizes whitespace and case", () => {
    const text = buildSongSearchText({
      title: "  Hello   World  ",
      sections: [
        {
          label: "Verse",
          lines: [{ lyrics: "Line   one", chords: [] }],
        },
      ],
    });

    expect(text).toBe("hello world line one");
  });

  it("truncates at SONG_INDEX_SEARCH_TEXT_MAX", () => {
    const longLyric = "word ".repeat(SONG_INDEX_SEARCH_TEXT_MAX);
    const text = buildSongSearchText({
      title: "Long Song",
      sections: [
        {
          label: "Verse",
          lines: [{ lyrics: longLyric, chords: [] }],
        },
      ],
    });

    expect(text.length).toBe(SONG_INDEX_SEARCH_TEXT_MAX);
  });
});

File: mobile/test/domain/song_search_text_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:lf_chords/domain/constants.dart';
import 'package:lf_chords/domain/song_search_text.dart';

void main() {
  const sampleSections = [
    {
      'label': 'Verse 1',
      'lines': [
        {'lyrics': 'Amazing grace, how sweet the sound', 'chords': []},
        {'lyrics': 'That saved a wretch like me', 'chords': []},
      ],
    },
  ];

  group('flattenSectionsLyrics', () {
    test('returns all lyric lines in order', () {
      expect(
        flattenSectionsLyrics(sampleSections),
        [
          'Amazing grace, how sweet the sound',
          'That saved a wretch like me',
        ],
      );
    });
  });

  group('buildSongSearchText', () {
    test('includes title, artist, tags, and lyrics', () {
      final text = buildSongSearchText(
        title: 'Amazing Grace',
        artist: 'John Newton',
        tags: ['lang:english', 'worship'],
        sections: sampleSections,
      );

      expect(text, contains('amazing grace'));
      expect(text, contains('john newton'));
      expect(text, contains('lang:english'));
      expect(text, contains('sweet the sound'));
      expect(text, contains('wretch like me'));
    });

    test('normalizes whitespace and case', () {
      final text = buildSongSearchText(
        title: '  Hello   World  ',
        sections: [
          {
            'label': 'Verse',
            'lines': [
              {'lyrics': 'Line   one', 'chords': []},
            ],
          },
        ],
      );

      expect(text, 'hello world line one');
    });

    test('truncates at songIndexSearchTextMax', () {
      final longLyric = List.filled(songIndexSearchTextMax, 'word ').join();
      final text = buildSongSearchText(
        title: 'Long Song',
        sections: [
          {
            'label': 'Verse',
            'lines': [
              {'lyrics': longLyric, 'chords': []},
            ],
          },
        ],
      );

      expect(text.length, songIndexSearchTextMax);
    });
  });
}
